File: library/ilib/string-cheese/ffi/string-cheese_ffi.py
# AC ilib: string-cheese — Python FFI
# Inlined by AC->PY compiler when "use ilib string-cheese" is declared.
import re as _re

_WS = " \t\n\r"  # \ws sentinel resolved to this

def _ws(s):
    return s if s != " \t\n\r" else None  # None → Python default strip (all whitespace)

# ── stringm functions ─────────────────────────────────────────────────────────
# b / f / t are the three string-prefix constructors (like Python b"" / f"" / t"").
# f-string and t-string do their interpolation at the compiler/IR level; the runtime
# helpers are passthroughs (mirrors ac_stringm_format in string_cheese_c.h).

def stringm_f(s, *args):
    return str(s)                 # f-string (formatted) — compiler interpolates {} at IR level

def stringm_t(s, *args):
    return str(s)                 # t-string (template, PEP 750) — template resolved at IR level

def stringm_b(s):
    return str(s).encode('utf-8')

def stringm_endian(b, order='little'):
    # Interpret a byte string as an unsigned integer ('little' or 'big' endian).
    if isinstance(b, str):
        b = b.encode('utf-8')
    return int.from_bytes(b, 'little' if str(order).lower().startswith('l') else 'big')

def stringm_upper(s):
    return str(s).upper()

def stringm_lower(s):
    return str(s).lower()

def stringm_find(s, pattern):
    p = str(pattern)
    if p == _WS:
        m = _re.search(r'\s', str(s))
        return m.start() if m else -1
    return str(s).find(p)

def stringm_strip(s, chars=None):
    if chars is None or chars == _WS:
        return str(s).strip()
    return str(s).strip(str(chars))

def stringm_replace(s, old, new):
    if str(old) == _WS:
        return _re.sub(r'\s+', str(new), str(s))
    return str(s).replace(str(old), str(new))

def stringm_split(s, sep=None):
    if sep is None or sep == _WS:
        return str(s).split()
    return str(s).split(str(sep))

def stringm_join(sep, parts):
    return str(sep).join(str(p) for p in parts)

def stringm_len(s):
    return len(str(s))

def stringm_startswith(s, prefix):
    return str(s).startswith(str(prefix))

def stringm_endswith(s, suffix):
    return str(s).endswith(str(suffix))

def stringm_count(s, sub):
    if str(sub) == _WS:
        return sum(1 for c in str(s) if c in _WS)
    return str(s).count(str(sub))

def stringm_format(template, **kwargs):
    return str(template).format(**kwargs)

class stringm:
    f          = staticmethod(stringm_f)
    t          = staticmethod(stringm_t)
    b          = staticmethod(stringm_b)
    endian     = staticmethod(stringm_endian)
    upper      = staticmethod(stringm_upper)
    lower      = staticmethod(stringm_lower)
    find       = staticmethod(stringm_find)
    strip      = staticmethod(stringm_strip)
    replace    = staticmethod(stringm_replace)
    split      = staticmethod(stringm_split)
    join       = staticmethod(stringm_join)
    length     = staticmethod(stringm_len)
    startswith = staticmethod(stringm_startswith)
    endswith   = staticmethod(stringm_endswith)
    count      = staticmethod(stringm_count)
    format     = staticmethod(stringm_format)

File: library/ilib/string-cheese/ffi/string-cheese_ffi.java
// AC ilib: string-cheese — Java implementation
import java.util.*;
import java.util.regex.*;

class StringCheese {

    private static final String _WS = " \t\n\r";

    private static boolean isWsSentinel(String s) {
        return s.equals(_WS);
    }

    /**
     * stringm_lower - Convert to lowercase
     */
    public static String lower(String s) {
        return s == null ? "" : s.toLowerCase();
    }

    /**
     * stringm_upper - Convert to uppercase
     */
    public static String upper(String s) {
        return s == null ? "" : s.toUpperCase();
    }

    /**
     * stringm_trim - Trim whitespace
     */
    public static String trim(String s) {
        return s == null ? "" : s.trim();
    }

    /**
     * stringm_strip - Strip characters from ends
     */
    public static String strip(String s, String chars) {
        if (s == null) return "";
        if (chars == null || isWsSentinel(chars)) {
            return s.trim();
        }

        int start = 0, end = s.length();
        while (start < end && chars.indexOf(s.charAt(start)) >= 0) start++;
        while (end > start && chars.indexOf(s.charAt(end - 1)) >= 0) end--;

        return s.substring(start, end);
    }

    /**
     * stringm_find - Find pattern position
     */
    public static long find(String s, String pattern) {
        if (s == null || pattern == null) return -1;

        if (isWsSentinel(pattern)) {
            for (int i = 0; i < s.length(); i++) {
                if (Character.isWhitespace(s.charAt(i))) {
                    return i;
                }
            }
            return -1;
        }

        int idx = s.indexOf(pattern);
        return idx >= 0 ? idx : -1;
    }

    /**
     * stringm_replace - Replace pattern
     */
    public static String replace(String s, String old, String newStr) {
        if (s == null || old == null) return s == null ? "" : s;

        if (isWsSentinel(old)) {
            String[] parts = s.trim().split("\\s+");
            return String.join(newStr == null ? " " : newStr, parts);
        }

        return s.replace(old, newStr == null ? "" : newStr);
    }

    /**
     * stringm_split - Split string
     */
    public static String[] split(String s, String sep) {
        if (s == null) return new String[]{};

        if (sep == null || isWsSentinel(sep)) {
            return s.trim().split("\\s+");
        }

        return s.split(Pattern.quote(sep), -1);
    }

    /**
     * stringm_split_nth - Get nth split part
     */
    public static String splitNth(String s, String sep, int n) {
        String[] parts = split(s, sep);
        if (n >= 0 && n < parts.length) {
            return parts[n];
        }
        return "";
    }

    /**
     * stringm_len - String length
     */
    public static long len(String s) {
        return s == null ? 0 : s.length();
    }

    /**
     * stringm_startswith - Check prefix
     */
    public static boolean startsWith(String s, String prefix) {
        return s != null && prefix != null && s.startsWith(prefix);
    }

    /**
     * stringm_endswith - Check suffix
     */
    public static boolean endsWith(String s, String suffix) {
        return s != null && suffix != null && s.endsWith(suffix);
    }

    /**
     * stringm_count - Count occurrences
     */
    public static long count(String s, String sub) {
        if (s == null || sub == null) return 0;

        if (isWsSentinel(sub)) {
            int cnt = 0;
            for (char c : s.toCharArray()) {
                if (Character.isWhitespace(c)) cnt++;
            }
            return cnt;
        }

        int count = 0;
        int idx = 0;
        while ((idx = s.indexOf(sub, idx)) >= 0) {
            count++;
            idx += sub.length();
        }
        return count;
    }

    /**
     * stringm_join - Join strings
     */
    public static String join(String sep, String... parts) {
        if (sep == null) sep = "";
        return String.join(sep, parts);
    }

    /**
     * stringm_format - Format string (passthrough)
     */
    public static String format(String t) {
        return t == null ? "" : t;
    }

    /**
     * stringm_ischar - Check if all alphabetic
     */
    public static boolean isChar(String s) {
        if (s == null || s.isEmpty()) return false;
        for (char c : s.toCharArray()) {
            if (!Character.isLetter(c)) return false;
        }
        return true;
    }

    /**
     * stringm_isws - Check if all whitespace
     */
    public static boolean isWS(String s) {
        if (s == null) return false;
        if (s.isEmpty()) return true;
        for (char c : s.toCharArray()) {
            if (!Character.isWhitespace(c)) return false;
        }
        return true;
    }

    /**
     * stringm_getline - Read line from stdin (stub)
     */
    public static String getline() {
        try {
            Scanner sc = new Scanner(System.in);
            return sc.hasNextLine() ? sc.nextLine() : "";
        } catch (Exception e) {
            return "";
        }
    }

    /**
     * stringm_scan - Scan stdin for pattern (stub)
     */
    public static boolean scan(String needle) {
        try {
            Scanner sc = new Scanner(System.in);
            while (sc.hasNextLine()) {
                if (sc.nextLine().contains(needle)) {
                    return true;
                }
            }
        } catch (Exception e) {
        }
        return false;
    }
}

// stringm namespace class — AC-generated Java uses stringm.upper(s), etc. Missing entirely
// before (verified: examples/stringm_demo.ac, "cannot find symbol: variable stringm" — every
// other backend resolves `stringm.*` through a native object/namespace of some kind; Java had
// the real StringCheese implementation but no dispatcher exposing it under the AC-level name).
// Names/arities mirror string-cheese.acl's stringm:* -> stringm.* mapping.
class stringm {
    public static String b(String s)                       { return StringCheese.trim(s); }
    public static String upper(String s)                   { return StringCheese.upper(s); }
    public static String lower(String s)                   { return StringCheese.lower(s); }
    public static long   find(String s, String pattern)    { return StringCheese.find(s, pattern); }
    public static String strip(String s, String chars)     { return StringCheese.strip(s, chars); }
    public static String strip(String s)                   { return StringCheese.strip(s, " \t\n\r"); }
    public static String replace(String s, String o, String n) { return StringCheese.replace(s, o, n); }
    public static String[] split(String s, String sep)     { return StringCheese.split(s, sep); }
    public static String splitNth(String s, String sep, int n) { return StringCheese.splitNth(s, sep, n); }
    public static String join(String sep, String... parts) { return StringCheese.join(sep, parts); }
    public static long   length(String s)                  { return StringCheese.len(s); }
    public static boolean startswith(String s, String p)   { return StringCheese.startsWith(s, p); }
    public static boolean endswith(String s, String p)     { return StringCheese.endsWith(s, p); }
    public static long   count(String s, String sub)       { return StringCheese.count(s, sub); }
    public static String format(String t)                  { return StringCheese.format(t); }
    public static String getline()                         { return StringCheese.getline(); }
    public static boolean scan(String needle)               { return StringCheese.scan(needle); }
}
